aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -1,57 +1,127 @@
 #include "ATen/cuda/CUDAStream.h"
 #include "ATen/cuda/CUDAContext.h"
 #include "ATen/cuda/Exceptions.h"
+#include "ATen/DeviceGuard.h"
 #include "ATen/Error.h"
 
 #include <mutex>
 #include <atomic>
+#include <cstdint>
+#include <deque>
+#include <vector>
 
 // Internal implementation is entirely hidden
+// Note: CUDAStreamInternals doubles for a THCStream
 struct CUDAStreamInternals {
-  bool is_destructible;
-  std::atomic<int> refcount;
-  int64_t device; // Note: cudaGetDevice works with int32_t, not int64_t
-  cudaStream_t stream;
+  CUDAStreamInternals() = default;
+
+  ~CUDAStreamInternals() {
+    if (stream) cudaStreamDestroy(stream);
+  }
```

Comment:
Doesn't have to be this PR, but it would be nice if `CUDAStreamInternals` stored `cudaStream_t` as a `std::unique_ptr<CUStream_st, CUDAStreamDeleter>` where CUDAStreamDeleter is a functor that calls `cudaStreamDestroy`. Example of this in the cudnn descriptor code.